### PR TITLE
Lookup modules located next to the file supplied on CLI

### LINKED ISSUE #426: [BUG] Cannot find source file for module...
Spec/Model at https://github.com/tlaplus/Examples/tree/master/specifications/ewd998

```shell
markus@avocado:/tmp/apalache(unstable)$ bin/apalache-mc check /home/markus/src/TLA/_specs/examples/specifications/ewd998/EWD998.tla
# Tool home: /tmp/apalache
# Package:   /tmp/apalache/mod-distribution/target/apalache-pkg-0.8.1-SNAPSHOT-full.jar
# JVM args: -Xmx4096m -DTLA-Library=/tmp/apalache/src/tla
#
# APALACHE version 0.8.1-SNAPSHOT build v0.7.2-117-g9e40275
#
# WARNING: This tool is in the experimental stage.
#          Please report bugs at: [https://github.com/informalsystems/apalache/issues]
# 
# Usage statistics is ON. Thank you!
# If you have changed your mind, disable the statistics with config --enable-stats=false.

Checker options: filename=/home/markus/src/TLA/_specs/examples/specifications/ewd998/EWD998.tla, init=, next=, inv= I@09:54:04.080
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/tmp/apalache/mod-distribution/target/apalache-pkg-0.8.1-SNAPSHOT-full.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
PASS #0: SanyParser                                               I@09:54:04.584
Parsing file /home/markus/src/TLA/_specs/examples/specifications/ewd998/EWD998.tla
Parsing file /tmp/Integers.tla
Parsing file /tmp/FiniteSets.tla
File does not exist: /tmp/apalache/src/tla/Utils.tla while looking in these directories: /tmp/apalache/src/tla/, jar:file:/tmp/apalache/mod-distribution/target/apalache-pkg-0.8.1-SNAPSHOT-full.jar!/tla2sany/StandardModules/
Error by TLA+ parser: *** Abort messages: 1

Unknown location

Cannot find source file for module Utils imported in module EWD998.


 E@09:54:04.818
It took me 0 days  0 hours  0 min  0 sec                          I@09:54:04.819
Total time: 0.741 sec                                             I@09:54:04.819
EXITCODE: ERROR (99)
```

```shell
markus@avocado:/tmp/apalache(unstable)$ /opt/toolbox/tla2tools.jar -deadlock /home/markus/src/TLA/_specs/examples/specifications/ewd998/EWD998
TLC2 Version 2.15 of Day Month 20?? (rev: 70a5e5d)
Warning: Please run the Java VM, which executes TLC with a throughput optimized garbage collector, by passing the "-XX:+UseParallelGC" property.
(Use the -nowarning option to disable this warning.)
Running breadth-first search Model-Checking with fp 58 and seed -6008330535915715549 with 1 worker on 4 cores with 5952MB heap and 64MB offheap memory [pid: 691579] (Linux 5.4.0-58-generic amd64, Ubuntu 11.0.9.1 x86_64, MSBDiskFPSet, DiskStateQueue).
Parsing file /home/markus/src/TLA/_specs/examples/specifications/ewd998/EWD998.tla
Parsing file /tmp/Integers.tla
Parsing file /tmp/FiniteSets.tla
Parsing file /home/markus/src/TLA/_specs/examples/specifications/ewd998/Utils.tla
Parsing file /tmp/Naturals.tla
Parsing file /tmp/Sequences.tla
Semantic processing of module Naturals
Semantic processing of module Integers
Semantic processing of module Sequences
Semantic processing of module FiniteSets
Semantic processing of module Utils
Semantic processing of module EWD998
Starting... (2021-01-11 09:53:09)
Computing initial states...
Computed 2 initial states...
Computed 4 initial states...
Computed 8 initial states...
Computed 16 initial states...
Computed 32 initial states...
Finished computing initial states: 64 distinct states generated at 2021-01-11 09:53:10.
Progress(9) at 2021-01-11 09:53:13: 415,711 states generated (415,711 s/min), 76,399 distinct states found (76,399 ds/min), 28,126 states left on queue.
Model checking completed. No error has been found.
  Estimates of the probability that TLC did not check all reachable states
  because two distinct states had the same fingerprint:
  calculated (optimistic):  val = 4.9E-7
  based on the actual fingerprints:  val = 1.4E-10
8707823 states generated, 1202278 distinct states found, 0 states left on queue.
The depth of the complete state graph search is 60.
The average outdegree of the complete state graph is 1 (minimum is 0, the maximum 9 and the 95th percentile is 3).
Finished in 42s at (2021-01-11 09:53:51)
```

### LINKED ISSUE #1477: Apalache fails to find module file
## Description

Let's assume following layout:

`.` <--- work directory where `java -jar apalache.jar ...` is called
`./dir/M0.tla` <-- Main module
`./dir/M1.tla` <-- `M0` extends `M1`

Then Apalache cannot find `./dir/M1.tla`. TLC works as expected.

## Input specification

```
---- MODULE M0 ----
EXTENDS M1
VARIABLE
    \* @type: Bool;
    x
Init == x = TRUE
Next == x' = ~x
Spec == Init /\ [][Next]_x
Inv == x => ~y
====
```

```
---- MODULE M1 ----
\* @type: Bool;
ap_0 == TRUE
====
```

## The command line parameters used to run the tool

`java -jar ../tools/apalache-pkg-0.20.2-full.jar check dir/M0.tla`

## Expected behavior

Apalache is able to find all tla-files

## Log files

```
# APALACHE
# version: 0.20.2
# build  : 5b59b10
#
# Usage statistics is OFF. We care about your privacy.
# If you want to help our project, consider enabling statistics with config --enable-stats=true.

Output directory: /mnt/c/work/TLA/tlc-qual/bug/_apalache-out/M0.tla/2022-03-12T21-03-30_5656217929895998894
Checker options: filename=dir/M0.tla, init=, next=, inv=          I@21:03:31.116
Tuning:                                                           I@21:03:31.122
PASS #0: SanyParser                                               I@21:03:31.124
Parsing file /mnt/c/work/TLA/tlc-qual/bug/dir/M0.tla
File does not exist: /mnt/c/work/TLA/tlc-qual/bug/M1.tla while looking in these directories: jar:file:/mnt/c/work/TLA/tlc-qual/tools/apalache-pkg-0.20.2-full.jar!/tla2sany/StandardModules/
Error by TLA+ parser: *** Abort messages: 1

Unknown location

Cannot find source file for module M1 imported in module M0.


 E@21:03:31.306
It took me 0 days  0 hours  0 min  0 sec                          I@21:03:31.307
Total time: 0.995 sec                                             I@21:03:31.308
EXITCODE: ERROR (255)
```

## System information

- Apalache version [`apalache-mc version`]: 0.20.2
- OS [e.g. Ubuntu Linux or Mac OS, and the current version]: Ubuntu 20.04
- JDK version [e.g. OpenJDK 0.8.0]:  openjdk version "14.0.2"

### PR DESCRIPTION
**Pull request by thpani**
Look up modules in the directory containing the main TLA+ file supplied on CLI. This was requested multiple times #426, #1477 and is what TLC does for module lookup.
Drop support for lookup in `TLA-Library`, unless the main file has no parent directory.
Drop support for `TLA_PATH`, since it becomes virtually useless.

This mirrors the behavior of TLC since https://github.com/tlaplus/tlaplus/commit/b905a86e31714bf5002290141b51b41ffe280c8e.

If we need `TLA_PATH`-like functionality down the road, we should parse the env variable in Apalache, then pass it into the filename resolver `SanyNameToStream` as additional library path.

The only remaining use of `TLA-Library` is to wire in the Apalache standard library in `src/tla/` for unit tests.

Fixes #426. Fixes #1477.

### CODE REVIEW

**Comment by shonfeder** [comment]
A few preliminary questions/notes based on the PR description:

1. Isn't the `TLA_PATH` still very useful if you want to add additional modules as a prelude or alternative "standard library"? 
2. I don't understand the rationale for 

   >  Drop support for lookup in TLA-Library, unless the main file has no parent directory.

   What is the relevance of the parent directory for the main file? And wouldn't this only be possible if the main file was at the root your file system?

3. I think we should definitely have integration tests for this kind of feature. Iiuc, since you've currently disabled the TLA_PATH, you're getting unintentional integration test coverage from https://github.com/informalsystems/apalache/blob/eae440964944ba842f1610a78444d2ff30ff47b4/test/tla/cli-integration-tests.md#L1177-L1185

**Comment by thpani** [comment]
> 1. Isn't the `TLA_PATH` still very useful if you want to add additional modules as a prelude or alternative "standard library"?

I don't know. Supporting it would mean to diverge from how TLC does module lookup, which may have unexpected results. I think we should remain compatible on this point. If someone still has a use case for `TLA_PATH`, it should be subsequently added to both Apalache and TLC.

> 2. I don't understand the rationale for
>    > Drop support for lookup in TLA-Library, unless the main file has no parent directory.
>    
>    
>    What is the relevance of the parent directory for the main file? And wouldn't this only be possible if the main file was at the root your file system?

To clarify, by parent directory I meant `dirname $tlafile`. So even for a main file in root, it would search `/`. I don't know the significance of this, but that's how it's implemented in TLC and SANY.

> 3. I think we should definitely have integration tests for this kind of feature.

Fair enough, I'm going to add integration tests. Just be warned that it'll add some non-trivial directory structure in `./test/tla`.

File: test/tla/cli-integration-tests.md
Comment (konnov) [inline test/tla/cli-integration-tests.md:2671]: Is this also the behavior of TLC? I find it rather confusing.

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala
Comment (konnov) [inline tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala:30]: I find this name confusing. Shall we call it like `useRelativePathLookup`?

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala
Comment (konnov) [inline tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala:46]: Is it a fallback case? I guess, the only case in Linux/MacOS is when the user passes `/` as the filename.

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala
Comment (konnov) [inline tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala:11]:
```suggestion
 *   Igor Konnov, Thomas Pani
```

**Review by konnov — APPROVED** [review]
Looks great! Thanks for taking care of this. If SANY removes support for `TLA-Library` in the future, we will have to revisit this resolver.

File: UNRELEASED.md
Comment (shonfeder) [inline UNRELEASED.md:15]: Perhaps the support for automatically finding tla modules in the file's directory should also be noted as a new feature?

File: docs/src/apalache/running.md
Comment (shonfeder) [inline docs/src/apalache/running.md:233]: Perhaps we can remove the mention of `APALACHE_HOME`, since we now include these resources in the fat jar distribution?

```suggestion
1. A small Apalache standard library
```

**Review by shonfeder — APPROVED** [review]

File: test/tla/module-lookup/Apalache.tla
Comment (thpani) [reply test/tla/module-lookup/Apalache.tla:1]: This is intentionally named `Apalache` to check that it takes precedence over the standard library module.
We can either not test this at all, or use another standard lib module name if you prefer? If the latter, what would be a good choice?

FYI @shonfeder

File: test/tla/cli-integration-tests.md
Comment (thpani) [reply test/tla/cli-integration-tests.md:2671]: Yes, I also tried with TLC. I find it confusing as well.
Ideally, it should just resolve within the same directory, then lookup standard modules. WDYT?

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala:46]: For the rare case of a non-hierarchical file system? In any case, `getParent()`

> Returns [...] null if this pathname does not name a parent directory.

so we have to handle it somehow.

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala:11]: Ah, are we keeping track of these? I've ignored them so far 😄

File: docs/src/apalache/running.md
Comment (thpani) [reply docs/src/apalache/running.md:233]: I think it's still helpful if people know what this "Apalache standard library" thing is?

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala:30]: Aaaah, I was looking for a good name 😄 Thanks!
Changed in 6e63be6f

File: UNRELEASED.md
Comment (thpani) [reply UNRELEASED.md]: Good point, updated in 180492be

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala
Comment (thpani) [reply tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala:46]: I just realized that `getParent()` also returns `null` on a relative path without parent directory.
This is called in 3 different places, so I'm calling `getAbsoluteFile()` in 168c3d88 below to make sure this is an absolute path.

File: tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala
Comment (konnov) [reply tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyNameToStream.scala]: Whenever I see them, I ask people to add them as co-authors :)

**Comment by konnov** [comment]
@andrey-kuprianov, we are removing support for `TLA_PATH`. I hope the MBT tools are not relying on that.

**Comment by shonfeder** [comment]
Good catch there asking MBT, Igor. Let's be sure to wait on their confirmation before merging this in.

**Review by shonfeder — APPROVED** [review]
LGTM!

File: test/tla/module-lookup/Apalache.tla
Comment (thpani) [reply test/tla/module-lookup/Apalache.tla]: As discussed offline, I have included a separate dummy module in [d3f95dd](https://github.com/informalsystems/apalache/pull/1491/commits/d3f95dda1061bf7b6e2d12cefd9f5c43fe2f373e)

**Pull request merged**
